plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -1,11 +1,247 @@
+
+    eval :: [Term] -> [String] -> MakeKnownM Term
+    eval args0 argStrings = case meaning of
+      BuiltinMeaning _ _ runtime -> go (_broRuntimeScheme runtime) (_broImmediateF runtime) args0
+      where
+        go ::
+          forall n.
+          RuntimeScheme n ->
+          ToRuntimeDenotationType Term n ->
+          [Term] ->
+          MakeKnownM Term
+        go sch f args = case (sch, args) of
+          (RuntimeSchemeArrow sch', a : as) -> do
+            res <- liftEither (f a)
+            go sch' res as
+          (RuntimeSchemeResult, []) -> f
+          (RuntimeSchemeAll sch', _) -> go sch' f args
+          _ -> error $ "Wrong number of args for builtin " <> show bn <> ": " <> show argStrings
+
+-- | Generate arguments to a builtin function
+genArgs :: DefaultFun -> Gen [(Term, String)]
+genArgs bn = case bn of
+  -- These functions are partial, so we manually generate their arguments.
+  IndexByteString -> do
+    s <- Gen.utf8 (Range.linear 5 100) Gen.enumBounded
+    i :: Integer <- fromIntegral <$> Gen.int (Range.linear 0 (BS.length s - 1))
+    pure [mkArg s, mkArg i]
+  HeadList -> pure . pure . mkArg =<< Gen.list (Range.linear 1 100) genInteger
+  TailList -> genArgs HeadList
+  UnConstrData -> pure . pure . mkArg =<< genConstr 5
+  UnMapData -> pure . pure . mkArg =<< genMap 5
+  UnListData -> pure . pure . mkArg =<< genList 5
+  UnIData -> pure . pure . mkArg =<< genI
+  UnBData -> pure . pure . mkArg =<< genB
+  -- The rest are total functions and we generate the arguments based on their `TypeScheme`s.
+  _ -> sequenceA $ case meaning of
+    BuiltinMeaning tySch _ _ -> go tySch
+      where
+        go :: forall args res. TypeScheme Term args res -> [Gen (Term, String)]
+        go = \case
+          TypeSchemeResult    -> []
+          TypeSchemeArrow sch -> genArg (typeRep @(Head args)) : go sch
+          TypeSchemeAll _ sch -> go sch
+  where
+    meaning :: BuiltinMeaning Term (CostingPart DefaultUni DefaultFun)
+    meaning = toBuiltinMeaning bn
+
+-- | Generate one argument to a builtin function based on its `TypeRep`.
+genArg :: forall k (a :: k). TypeRep a -> Gen (Term, String)
+genArg tr
+  | Just HRefl <- eqTypeRep tr (typeRep @()) = pure $ mkArg ()
+  | Just HRefl <- eqTypeRep tr (typeRep @Integer) = pure . mkArg =<< genInteger
+  | Just HRefl <- eqTypeRep tr (typeRep @Int) = pure . mkArg =<< genInteger
+  | Just HRefl <- eqTypeRep tr (typeRep @Bool) = pure . mkArg =<< Gen.bool
+  | Just HRefl <- eqTypeRep tr (typeRep @BS.ByteString) = pure . mkArg =<< genByteString
+  | Just HRefl <- eqTypeRep tr (typeRep @Text) = pure . mkArg =<< genText
+  | Just HRefl <- eqTypeRep tr (typeRep @Data) = pure . mkArg =<< genData 5
+  | Just [SomeTypeRep tr1, SomeTypeRep tr2] <- matchTyCon @(,) tr = do
+      (arg1, argStr1) <- genArg tr1
+      (arg2, argStr2) <- genArg tr2
+      case (arg1, arg2) of
+        (Constant _ (Some (ValueOf uni1 val1)), Constant _ (Some (ValueOf uni2 val2))) ->
```

Comment:
Maybe this function should return a `Some (ValueOf uni)` instead of a `Term`, then you could omit at least some of this match.

Reply:
And then since it couldn't fail you could push it into the matches on lines 133 and 134.

Reply:
Sounds like a good idea!